pkg/skaffold/kubernetes/debugging/container_manager.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2019 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package debugging
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/debug"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/event"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+)
+
+var (
+	// For testing
+	aggregatePodWatcher = kubernetes.AggregatePodWatcher
+
+	notifyDebuggingContainerStarted    = event.DebuggingContainerStarted
+	notifyDebuggingContainerTerminated = event.DebuggingContainerTerminated
+)
+
+type ContainerManager struct {
+	output      io.Writer
+	cli         *kubectl.CLI
+	podSelector kubernetes.PodSelector
+	namespaces  []string
+	active      map[string]string // set of containers that have been notified
+	aggregate   chan watch.Event
+}
```

Comment:
I had originally hoped to merge this into the port-forwarding `ForwarderManager`, but since there are some language runtimes that don't actually use forwarded ports (they instead require `kubectl ssh` and operate across stdin/stdout), I introduced this new `ContainerManager`.  The lifecycle of a debugging container doesn't match forwarded ports either.

Reply:
Which language runtimes require stdin/stdout?

Reply:
.NET

Reply:
.NET